README.md
Compiler and Debugger for MSP430 Texas Instruments
==================================================

[![Compiler of MSP430](https://github.com/tecip-nes/contiki-tres/wiki/Building-the-latest-version-of-mspgcc)](https://github.com/tecip-nes/contiki-tres/wiki/Building-the-latest-version-of-mspgcc)

To build the latest version of mspgcc (which supports the MSP430X extension) you can use the following script (which is a modified version of the one you can find on the old Contiki wiki).

[![More information about Compiler](https://github.com/contiki-os/contiki/wiki/Setup-Contiki-Toolchain-in-Arch-Linux)](https://github.com/contiki-os/contiki/wiki/Setup-Contiki-Toolchain-in-Arch-Linux)

[![Debugger of MSP430](https://github.com/dlbeer/mspdebug)](https://github.com/dlbeer/mspdebug)

MSPDebug is a free debugger for use with MSP430 MCUs. It supports FET430UIF, eZ430, RF2500 and Olimex MSP430-JTAG-TINY programmers, as well as many other compatible devices. It can be used as a proxy for gdb or as an independent debugger with support for programming, disassembly and reverse engineering.

  * Userspace only: no kernel modifications required.
  * Works with RF2500, eZ430, FET430UIF (V2 and V3), Launchpad, Chronos,
    GoodFET, Olimex MSP430-JTAG-TINY and MSP430-JTAG-ISO programmers.
    Also supports the TI flash bootloader.
  * Can act as a GDB remote stub (replacement for msp430-gdbproxy)
    and/or a GDB client.
  * Can single-step, program, run to breakpoint and inspect memory on
    supported devices.
  * Can be used to access the FET430UIF bootloader.
  * Supports Intel HEX, ELF32, BSD symbol table, COFF, TI Text and
    SREC file formats.
  * Can disassemble code in memory, including translating addresses to
    symbols.
  * Includes reverse-engineering features such as instruction search,
    call-graph analysis and symbol table editing.
  * Simulation mode allows execution of MSP430 code without hardware.
  * Cross-platform: compiles on Linux, *BSD, OS/X and Windows.

Installation in Ubuntu/Linux
============================

To install the compiler for 64bits:

$ sudo cp -rf mspgcc-4.7.2 /opt/

To install the compiler for 32bits:

$ sudo install-mspgcc-4.7.2.sh

To install the debugger for 64 bits operation system.

$ cp mspdebug-64bits/mspdebug /usr/local/bin/ 

To install the debugger for 32 bits operation system.

$ cp mspdebug-32bits/mspdebug /usr/local/bin/ 

To support automatical permissions for ttyUSBx ports

$ sudo cp 99-usb-serial.rules  /etc/udev/rules.d/99-usb-serial.rules

Include the following text to ~/.bashrc
=======================================

function msp430-472 () {

    export PATH=/opt/mspgcc-4.7.2/bin:$PATH  # Experimental 20 bits.

    MSP430_GCC_VERSION=`msp430-gcc -dumpversion`

    if [ "$MSP430_GCC_VERSION" == "4.7.2" ]; then

        echo "Compiler msp430-gcc 4.7.2 for MSP430 Experimental 20bits"

    else

        echo "ERROR: Compiler msp430-gcc $MSP430_GCC_VERSION != 4.7.2 for MSP430 Experimental 20bits"

    fi

}

Usage 
=====

make                # Compile in Contiki

make app_pwm.upload0  TARGET=odin5438A    # Program ttyUSB0

make app_pwm.upload1  TARGET=odin5438A    # Program ttyUSB1

mspdebug -d /dev/ttyUSB0 debug      # Logging ttyUSB0

mspdebug -d /dev/ttyUSB1 debug      # Logging ttyUSB1

compilation-5438a.sh msp430x54x_pwm.c   # Compile and Program in ttyUSB0 using example-msp430f54xxA




mspdebug-0.21/ChangeLog
Modification 0.21 - Nov 2013
	* Add flash_bsl for msp430f5xxx.
	* Add reset for msp430f5xxx.
	* Add debug for msp430f5xxx.
	* Add tapslip for msp430f5xxx.

Version 0.21 - 26 Nov 2012
	* Support for flat binary files.
	* Bug fixes and refactoring of drivers.
	* Support for UIF BSL access via USB.
	* Support for Olimex MSP430-JTAG-ISO-MK2, including power
	  profiling.
	* Support for parallel JTAG.
	* Support for many new chips when using Olimex drivers.
	* Add new interface mode for use with front-ends.

Version 0.20 - 14 Aug 2012
	* Support for new chips: MSP430F6736, MSP430F2252, MSP430G2403,
	  MSP430F6435.
	* Support for building under Cygwin.
	* Support for watchpoints using MSP430.DLL (tilib) driver.
	* Support for GoodFET (memory access/programming only).
	* Various bug fixes, including many disassembly fixes and an
	  issue causing FRAM reset-vector corruption.
	* Basic C++ demangling support.
	* New commands: "fill" and "verify".
	* Erase requests are now ignored when using FRAM chips.

Version 0.19 - 3 Mar 2012
	* Support for MSP430-JTAG-ISO on platforms other than Linux, via
	  both serial drivers and raw USB access.
	* Support for new chips: MSP430F5342, MSP430F5329, MSP430F2418.
	* Fixed support for later-model MSP430-JTAG-TINY rev 2.
	* Fixed build warnings on many systems.
	* Fixed output buffering to allow correct interaction with
	  Eclipse plugins.

Version 0.18 - 8 Nov 2011
	* Support for new chips: MSP430F2121, MSP430F2012, MSP430F449.
	* Support for raw USB access to FET430UIF and eZ430-F2013.
	* Support for TI MSP430 library.
	* Various Win32 bug fixes.
	* GDB protocol bug fixes.
	* Section names are displayed when programming.

Version 0.17 - 24 Aug 2011
	* Support for new chips: MSP430FR5739, MSP430F5418,
	  MSP430AFE253, MSP430F5527, MSP430F5510, MSP430F5528.
	* Added -s option to distinguish programming tools by serial
	  number.
	* Ported to Win32 (can be compiled with MinGW).
	* Bug fixes for GDB stub interface and simulator.
	* Reader now supports tilde-expansion and ^[# key.
	* Better diagnostics for FET devices (more error codes).

Version 0.16 - 14 Jun 2011
	* Support for new chips: MSP430G2252, MSP430G2553, MSP430F235,
	  MSP430F427.
	* Added GDB client driver.
	* Added support for Olimex MSP-JTAG-TINY rev 2.
	* Bug fix in flash-bsl driver for large memory transfers.

Version 0.15 - 5 Apr 2011
	* Support for new chips: MSP430G2452, MSP430F5525,
	  MSP430F47173, MSP430F5526.
	* Added --force-reset option.
	* Added "alias" command.
	* IO peripheral simulator (interrupts, hardware multiplier,
	  GPIO, Timer_A, Watchdog Timer+).
	* Configurable block size for FET memory transfers.
	* Fix build problems on Cygwin and OpenBSD.
	* Various corrections to manual page.

Version 0.14 - 7 Feb 2011
	* Corrected email address in AUTHORS file.
	* Support for new chips: MSP430F2272, MSP430F2122.
	* Support for different erase types with flash_bsl driver.
	* Fixed build problems on OS/X and OpenBSD.
	* Bug fix: incorrect simulation of SUB/SUBC/CMP opcodes.
	* Bug fix: faulty response to gdb "read registers" command.

Version 0.13 - 9 Nov 2010
	* Bug fix: incorrect handling of Intel HEX segment offsets.
	* Support for new chips: MSP430F1121, MSP430F2131, MSP430F2617,
	  MSP430F247, MSP430F4270, MSP430G2231, CC430F5133.
	* Support for Olimex MSP430-JTAG-ISO.
	* FET driver retries with a reset if the first attempt to
	  connect fails. This should fix the problem which requires
	  devices to be replugged after each session.
	* Notify the user when programming in quiet mode, and display
	  version banner unless starting in quiet mode.

Version 0.12 - 12 Oct 2010
	* Bug fixes: unaligned flash read/write, reduced memory block
	  size, MSP430F5438 support, BSL memory ranges.
	* Support for MSP430FG4618.
	* New commands: load, locka, exit.
	* Support for flash BSL (but see notes in manual).
	* Preliminary support for segment and mass erase (see "erase"
	  command).
	* Reset is now issued on disconnect, rather than on init.
	* Allow specification of default input radix (see "iradix"
	  option).
	* Allow partial unambiguous command matches.
	* Repeat execution of certain commands in reader.
	* Commands read from a file are now printed as they are
	  executed.

Version 0.11 - 9 Sep 2010
	* Support for programming and debugging 20-bit devices.
	* Support for MSP430X instruction set.
	* Support for new chips: MSP430F2234, CC430F5137, MSP430F2618,
	  MSP430F1612, MSP430F47197, MSP430F1232, MSP430F413,
	  MSP430F2370.
	* Bug fixes: RF2500 support on OS/X, chip identification using
	  FET.
	* Added quiet mode to supress debug output.
	* Faster gdb transfers.
	* Faster polling with FET driver (quicker single-stepping and
	  breakpoint detection).
	* Commands can now be issued, and output received, via the gdb
	  interface (using "monitor").
	* Support for FET430UIF firmware updates via BSL.

Version 0.10 - 4 Aug 2010
	* Bug fixes for gdb interface, FET driver.
	* Added support for MSP430F5529.
	* Added support for COFF, S19 and TI-Text file formats.
	* Support multiple breakpoints (see "setbreak", "delbreak" and
	  "break" commands).

Version 0.9 - 29 Jun 2010
	* Added support for MSP430F2132, MSP430F169.
	* Fixed support for FET430UIF JTAG with 5xxx series devices.
	* Added support for Olimex MSP-JTAG-TINY programmer.
	* Multiple instances of USB programmers can now be used
	  simultaneously.
	* Cleaner command line options (see "--help" for information).

Version 0.8 - 22 May 2010
	* Bug fixes for disassembly, chip erase, gdb interface,
	  simulation.
	* Modifications to improve portability to Cygwin and Mac OS X.
	* Implemented call graph analysis (see "cgraph" command).
	* Better FET protocol support for firmware versions >= 20300000.

Version 0.7 - 30 Apr 2010
	* Symbol tables can now be edited and exported (see "sym"
	  command).
	* Bug fixes for disassembly.
	* Added option for coloured output.
	* Can search memory for instructions matching a pattern (see
	  "isearch" command).

mspdebug-0.21/ui/main.c
/* MSPDebug - debugging tool for MSP430 MCUs
 * Copyright (C) 2009-2012 Daniel Beer
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include "dis.h"
#include "device.h"
#include "binfile.h"
#include "stab.h"
#include "util.h"
#include "usbutil.h"
#include "gdb.h"
#include "rtools.h"
#include "sym.h"
#include "devcmd.h"
#include "expr.h"
#include "opdb.h"
#include "reader.h"
#include "output.h"
#include "output_util.h"
#include "simio.h"
#include "ctrlc.h"

#include "sim.h"
#include "bsl.h"
#include "fet.h"
#include "vector.h"
#include "fet_db.h"
#include "fet_olimex_db.h"
#include "gdbc.h"
#include "tilib.h"
#include "goodfet.h"
#include "input.h"
#include "input_async.h"
#include "pif.h"

#include "flash_bsl.h"
#include "reset.h"
#include "debug.h"
#include "tapslip.h"

#define OPT_NO_RC		0x01
#define OPT_EMBEDDED		0x02

struct cmdline_args {
	const char		*driver_name;
	int			flags;
	struct device_args	devarg;
};

static const struct device_class *const driver_table[] = {
	&device_rf2500,
	&device_olimex,
	&device_olimex_v1,
	&device_olimex_iso,
	&device_olimex_iso_mk2,
	&device_sim,
	&device_uif,
	&device_bsl,
	&device_flash_bsl,
	&device_gdbc,
	&device_tilib,
	&device_goodfet,
	&device_pif,
	&device_reset,
	&device_debug,
	&device_tapslip,
};

static const char *version_text =
"MSPDebug version 0.21 - debugging tool for MSP430 MCUs\n"
"Copyright (C) 2009-2012 Daniel Beer <dlbeer@gmail.com>\n"
"This is free software; see the source for copying conditions.  There is NO\n"
"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR "
"PURPOSE.\n";

static void usage(const char *progname)
{
	int i;

	printc_err("Usage: %s [options] <driver> [command ...]\n"
"\n"
"    -q\n"
"        Start in quiet mode.\n"
"    -d device\n"
"        Connect via the given tty device, rather than USB.\n"
"    -U bus:dev\n"
"        Specify a particular USB device to connect to.\n"
"    -s serial\n"
"        Specify a particular device serial number to connect to.\n"
"    -j\n"
"        Use JTAG, rather than Spy-Bi-Wire (UIF devices only).\n"
"    -v voltage\n"
"        Set the supply voltage, in millivolts.\n"
"    -n\n"
"        Do not read ~/.mspdebug on startup.\n"
"    --long-password\n"
"        Send 32-byte IVT as BSL password (flash-bsl only)\n"
"    --help\n"
"        Show this help text.\n"
"    --fet-list\n"
"        Show a list of devices supported by the FET driver.\n"
"    --fet-force-id string\n"
"        Override the device ID returned by the FET.\n"
"    --usb-list\n"
"        Show a list of available USB devices.\n"
"    --force-reset\n"
"        Force target reset in initialization sequence.\n"
"    --allow-fw-update\n"
"        Update FET firmware (tilib only) if necessary.\n"
"    --require-fw-update <image.txt>\n"
"        Require FET firmware update. The required image format depends\n"
"        on the driver.\n"
"    --version\n"
"        Show copyright and version information.\n"
"    --embedded\n"
"        Run in embedded mode.\n"
"\n"
"Most drivers connect by default via USB, unless told otherwise via the\n"
"-d option. By default, the first USB device found is opened.\n"
"\n"
"If commands are given, they will be executed. Otherwise, an interactive\n"
"command reader is started.\n\n",
		progname);

	printc("Available drivers are:\n");
	for (i = 0; i < ARRAY_LEN(driver_table); i++) {
		const struct device_class *drv = driver_table[i];

		printc("    %s\n        %s\n", drv->name, drv->help);
	}
}

static void process_rc_file(void)
{
	const char *home = getenv("HOME");
	char text[256];

	if (!home)
		return;

	snprintf(text, sizeof(text), "%s/.mspdebug", home);
	if (!access(text, F_OK))
		process_file(text, 0);
}

static int add_fet_device(void *user_data, const struct fet_db_record *r)
{
	struct vector *v = (struct vector *)user_data;

	return vector_push(v, &r->name, 1);
}

static int add_fet_olimex_device(void *user_data, const char *name)
{
	struct vector *v = (struct vector *)user_data;

	return vector_push(v, &name, 1);
}

static int list_devices(void)
{
	struct vector v;

	vector_init(&v, sizeof(const char *));
	if (fet_db_enum(add_fet_device, &v) < 0) {
		pr_error("couldn't allocate memory");
		vector_destroy(&v);
		return -1;
	}

	printc("Devices supported by FET driver:\n");
	namelist_print(&v);
	vector_destroy(&v);

	vector_init(&v, sizeof(const char *));
	if (fet_olimex_db_enum(add_fet_olimex_device, &v) < 0) {
		pr_error("couldn't allocate memory");
		vector_destroy(&v);
		return -1;
	}

	printc("\n");
	printc("Devices supported by Olimex FET driver:\n");
	namelist_print(&v);
	vector_destroy(&v);

	return 0;
}

static int parse_cmdline_args(int argc, char **argv,
			      struct cmdline_args *args)
{
	int opt;
	static const struct option longopts[] = {
		{"help",                0, 0, 'H'},
		{"fet-list",            0, 0, 'L'},
		{"fet-force-id",        1, 0, 'F'},
		{"usb-list",            0, 0, 'I'},
		{"version",             0, 0, 'V'},
		{"long-password",       0, 0, 'P'},
		{"force-reset",		0, 0, 'R'},
		{"allow-fw-update",	0, 0, 'A'},
		{"require-fw-update",	1, 0, 'M'},
		{"embedded",		0, 0, 'E'},
		{NULL, 0, 0, 0}
	};
	int want_usb = 0;

	while ((opt = getopt_long(argc, argv, "d:jv:nU:s:q",
				  longopts, NULL)) >= 0)
		switch (opt) {
		case 'q':
			{
				static const union opdb_value v = {
					.boolean = 1
				};

				opdb_set("quiet", &v);
			}
			break;

		case 'E':
			args->flags |= OPT_EMBEDDED;
			break;

		case 'A':
			args->devarg.flags |= DEVICE_FLAG_DO_FWUPDATE;
			break;

		case 'I':
			usb_init();
			usb_find_busses();
			usb_find_devices();
			usbutil_list();
			exit(0);

		case 'd':
			args->devarg.path = optarg;
			args->devarg.flags |= DEVICE_FLAG_TTY;
			break;

		case 'M':
			args->devarg.require_fwupdate = optarg;
			break;

		case 'U':
			args->devarg.path = optarg;
			want_usb = 1;
			break;

		case 's':
			args->devarg.requested_serial = optarg;
			break;

		case 'L':
			exit(list_devices());

		case 'F':
			args->devarg.forced_chip_id = optarg;
			break;

		case 'H':
			usage(argv[0]);
			exit(0);

		case 'V':
			printc("%s", version_text);
			exit(0);

		case 'v':
			args->devarg.vcc_mv = atoi(optarg);
			break;

		case 'j':
			args->devarg.flags |= DEVICE_FLAG_JTAG;
			break;

		case 'n':
			args->flags |= OPT_NO_RC;
			break;

		case 'P':
			args->devarg.flags |= DEVICE_FLAG_LONG_PW;
			break;

		case 'R':
			args->devarg.flags |= DEVICE_FLAG_FORCE_RESET;
			break;

		case '?':
			printc_err("Try --help for usage information.\n");
			return -1;
		}

	if (want_usb && (args->devarg.flags & DEVICE_FLAG_TTY)) {
		printc_err("You can't simultaneously specify a serial and "
			"a USB device.\n");
		return -1;
	}

	if (optind >= argc) {
		printc_err("You need to specify a driver. Try --help for "
			"a list.\n");
		return -1;
	}

	args->driver_name = argv[optind];
	optind++;

	return 0;
}

int setup_driver(struct cmdline_args *args)
{
	int i;

	i = 0;
	while (i < ARRAY_LEN(driver_table) &&
	       strcasecmp(driver_table[i]->name, args->driver_name))
		i++;
	if (i >= ARRAY_LEN(driver_table)) {
		printc_err("Unknown driver: %s. Try --help for a list.\n",
			args->driver_name);
		return -1;
	}

	if (stab_init() < 0)
		return -1;

	device_default = driver_table[i]->open(&args->devarg);
	if (!device_default) {
		stab_exit();
		return -1;
	}

	return 0;
}

#ifdef __Windows__
static int sockets_init(void)
{
	WSADATA data;

	if (WSAStartup(MAKEWORD(2, 2), &data)) {
		printc_err("Winsock init failed");
		return -1;
	}

	return 0;
}

static void sockets_exit(void)
{
	WSACleanup();
}
#else
static int sockets_init(void) { return 0; }
static void sockets_exit(void) { }
#endif

int main(int argc, char **argv)
{
	struct cmdline_args args = {0};
	int ret = 0;

	setvbuf(stderr, NULL, _IOFBF, 0);
	setvbuf(stdout, NULL, _IOFBF, 0);

	opdb_reset();
	ctrlc_init();

	args.devarg.vcc_mv = 3000;
	args.devarg.requested_serial = NULL;
	if (parse_cmdline_args(argc, argv, &args) < 0)
		goto fail_parse;

	if (args.flags & OPT_EMBEDDED)
		input_module = &input_async;
	if (input_module->init() < 0)
		goto fail_input;

	output_set_embedded(args.flags & OPT_EMBEDDED);

	if (sockets_init() < 0) {
		ret = -1;
		goto fail_sockets;
	}

	printc_dbg("%s\n", version_text);
	if (setup_driver(&args) < 0) {
		ret = -1;
		goto fail_driver;
	}

	if (device_probe_id(device_default) < 0)
		printc_err("warning: device ID probe failed\n");

	simio_init();

	if (!(args.flags & OPT_NO_RC))
		process_rc_file();

	/* Process commands */
	if (optind < argc) {
		while (optind < argc) {
			if (process_command(argv[optind++]) < 0) {
				ret = -1;
				break;
			}
		}
	} else {
		reader_loop();
	}

	simio_exit();
	device_destroy();
	stab_exit();
fail_driver:
	sockets_exit();
fail_sockets:
	input_module->exit();
fail_input:
fail_parse:

	/* We need to do this on Windows, because in embedded mode we
	 * may still have a running background thread for input. If so,
	 * returning from main() won't cause the process to terminate.
	 */
#if defined(__Windows__) || defined(__CYGWIN__)
	ExitProcess(ret);
#endif
	return ret;
}
